### ISSUE TITLE
Support Postgres DB storage

### ISSUE THREAD

**Issue by markphelps** [markphelps/flipt#62]
Re: #41 

Add support for external DB storage with Postgres.

## Requirements:

* Satisfy the same interface in the [storage](https://github.com/markphelps/flipt/blob/master/storage/storage.go) package.
* Be configurable, meaning the user can swap out the default SQLite implementation with Postgres via the config file. This will likely mean changing the config format for `db`, ex:
https://github.com/markphelps/flipt/blob/3d4f3f02f7634d88d0219a2f733f5298dd0fc7d5/config/default.yml#L18
As path doesn't make sense for a Postgres database, I'm thinking something like this for Postgres:

```
db:
  url: postgresql://localhost:5432/flipt
```

And changing the default SQLite config to look something like this:
```
db:
  url: sqlite3:///var/opt/flipt/flipt.db
```

Or something similar, still on the fence about what would work best here. Feedback would be greatly appreciated!
* Use `github.com/Masterminds/squirrel`
* Handle foreign key/constraint errors like: https://github.com/markphelps/flipt/blob/3d4f3f02f7634d88d0219a2f733f5298dd0fc7d5/storage/flag.go#L234

This may make sense to abstract away the error handling since that is what is most likely to be different between the different databases. We might be able to get away with not changing the actual query logic much at all since `squirrel` and Go's `sql` lib should do most of the heavy lifting.

**Comment by harunalfat** [comment]
Glad to know you plan to do this. I plan to use Flipt for my current employer project, but our current infrastructure force us to use minimum of two pods of a service, and with its sqlite implementation, its problematic when the data of the two pods is out of sync.

I'm glad if I can contribute in this feature. The migration is hardcoded in SQL, is it works in Postgres too?

**Comment by markphelps** [comment]
I haven't tried the migrations with Postgres yet, but I plan to this week.

I use [golang-migrate](https://github.com/golang-migrate/migrate), to handle the migrations and according to them, the migrations are 'dumb' and processed directly by the db driver: https://github.com/golang-migrate/migrate/blob/master/MIGRATIONS.md#migration-content-format

So I'm thinking of changing the migrations paths from `/config/migrations` to `/config/migrations/{db}`, so separate migrations for sqlite and postgres incase the syntax is different at some point in the future.

Another option is to use a different [source](https://github.com/golang-migrate/migrate#migration-sources) for the migrations such as GitHub or S3, instead of serving them directly from the filesystem.. although this would mean a dependency on network connection when starting up Flipt for the first time.

**Comment by markphelps** [comment]
I'd be curious to know how other projects handle multiple database backends and migrations. I think GitLab supports multiple engines? Will need to do some research on this.

**Comment by harunalfat** [comment]
I think using mysql/mariadb for GitLab is deprecated. It will works for most cases, but then have several limitations

**Comment by markphelps** [comment]
Will push up a WIP branch to start hacking on this tonight or tomorrow

**Comment by markphelps** [comment]
WIP PR here: #68

**Comment by markphelps** [comment]
@harunalfat Postgres support has landed in v0.5.0.

Please give it a try and let me know what you think or if I can help in any way. Thanks!
